src/client/Microsoft.Identity.Client/ManagedIdentity/Csr.cs

```diff
@@ -0,0 +1,439 @@
+        private static byte[] EncodeAsn1ObjectIdentifier(int[] oid)
+        {
+            if (oid == null || oid.Length < 2)
+                throw new ArgumentException("OID must have at least 2 components");
+
+            var bytes = new System.Collections.Generic.List<byte>();
+
+            // First two components are encoded as (first * 40 + second)
+            bytes.AddRange(EncodeOidComponent(oid[0] * 40 + oid[1]));
+
+            // Remaining components
+            for (int i = 2; i < oid.Length; i++)
+            {
+                bytes.AddRange(EncodeOidComponent(oid[i]));
+            }
+
+            return EncodeAsn1Tag(0x06, bytes.ToArray());
+        }
+
+        /// <summary>
+        /// Encodes an ASN.1 context-specific tag.
+        /// </summary>
+        private static byte[] EncodeAsn1ContextSpecific(int tagNumber, byte[] content)
+        {
+            byte tag = (byte)(0xA0 | tagNumber); // Context-specific, constructed
+            return EncodeAsn1Tag(tag, content);
+        }
+
+        /// <summary>
+        /// Encodes an ASN.1 tag with length and content.
+        /// </summary>
+        private static byte[] EncodeAsn1Tag(byte tag, byte[] content)
+        {
+            byte[] lengthBytes = EncodeAsn1Length(content.Length);
+            byte[] result = new byte[1 + lengthBytes.Length + content.Length];
+            result[0] = tag;
+            Array.Copy(lengthBytes, 0, result, 1, lengthBytes.Length);
+            Array.Copy(content, 0, result, 1 + lengthBytes.Length, content.Length);
+            return result;
+        }
+
+        /// <summary>
+        /// Encodes ASN.1 length field.
+        /// </summary>
+        private static byte[] EncodeAsn1Length(int length)
+        {
+            if (length < 0x80)
+            {
+                return new byte[] { (byte)length };
+            }
+
+            var lengthBytes = new System.Collections.Generic.List<byte>();
+            int temp = length;
+            while (temp > 0)
+            {
+                lengthBytes.Insert(0, (byte)(temp & 0xFF));
+                temp >>= 8;
+            }
+
+            byte[] result = new byte[lengthBytes.Count + 1];
+            result[0] = (byte)(0x80 | lengthBytes.Count);
+            lengthBytes.CopyTo(result, 1);
+            return result;
+        }
+
+        /// <summary>
+        /// Encodes a single OID component using variable-length encoding.
+        /// </summary>
+        private static byte[] EncodeOidComponent(int value)
+        {
+            if (value == 0)
+                return new byte[] { 0x00 };
+
+            var bytes = new System.Collections.Generic.List<byte>();
+            int temp = value;
+
+            bytes.Insert(0, (byte)(temp & 0x7F));
+            temp >>= 7;
+
+            while (temp > 0)
+            {
+                bytes.Insert(0, (byte)((temp & 0x7F) | 0x80));
+                temp >>= 7;
+            }
+
+            return bytes.ToArray();
+        }
```

Comment:
All of these seem like they can/should use System.Formats.Asn1.

Reply:
Looks like this is not available 462 and 472. Do you think it's worth it to conditionally use `System.Formats.Asn1` for net5.0+? Or would it be clearer to use the same code for all versions (how it is now).

Reply:
It's available, you just need to reference the System.Formats.Asn1 NuGet package.

Reply:
I've addressed this in #5449